src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirModel.cs

```diff
@@ -0,0 +1,200 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.SqlClient;
+using System.Linq;
+using EnsureThat;
+using Hl7.Fhir.Model;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Fhir.Core.Features.Definition;
+using Microsoft.Health.Fhir.SqlServer.Configs;
+using Newtonsoft.Json;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Storage
+{
+    /// <summary>
+    /// Maintains IDs for resource types, search parameters, systems, and codes for quantity search parameters.
+    /// There are typically on the order of tens or hundreds of distinct values for each of these, but are reused
+    /// many many times in the database. For more compact storage, we use IDs instead of the strings when referencing these.
+    /// Also, because the number of distinct values is small, we can maintain all values in memory and avoid joins when querying.
+    /// </summary>
+    public class SqlServerFhirModel
+    {
+        private readonly SqlServerDataStoreConfiguration _configuration;
+        private readonly ILogger<SqlServerFhirModel> _logger;
+        private readonly ISearchParameterDefinitionManager _searchParameterDefinitionManager;
+        private readonly Dictionary<string, short> _resourceTypeToId = new Dictionary<string, short>(StringComparer.Ordinal);
+        private readonly Dictionary<short, string> _resourceTypeIdToTypeName = new Dictionary<short, string>();
+        private readonly Dictionary<string, short> _searchParamUriToId = new Dictionary<string, short>(StringComparer.Ordinal);
+        private readonly ConcurrentDictionary<string, int> _systemToId = new ConcurrentDictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+        private readonly ConcurrentDictionary<string, int> _quantityCodeToId = new ConcurrentDictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+
+        public SqlServerFhirModel(SqlServerDataStoreConfiguration configuration, ISearchParameterDefinitionManager searchParameterDefinitionManager, ILogger<SqlServerFhirModel> logger)
+        {
+            EnsureArg.IsNotNull(configuration, nameof(configuration));
+            EnsureArg.IsNotNull(searchParameterDefinitionManager, nameof(searchParameterDefinitionManager));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _configuration = configuration;
+            _logger = logger;
+            _searchParameterDefinitionManager = searchParameterDefinitionManager;
+            Initialize();
```

Comment:
There doesn't appear to be any code that makes sure that the schema initialization has occurred before this is called.

Reply:
Added lazy initialization to the `SqlServerFhirModel` class, which can begin once a current schema version is known.